### PR TITLE
JAMES-2244 Relax $Draft condition

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 1]
```diff
@@ -1470,7 +1467,7 @@ public void setMessagesShouldMarkAsDraftWhenIsDraftPassed() {
     }
 
     @Test
-    public void setMessagesShouldRejectCreateInDraftAndOutboxForASingleMessage() {
+    public void setMessagesShouldNotRejectCreateInDraftAndOutboxForASingleMessage() {
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: how is it possible ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature [hunk 2]
```diff
@@ -135,16 +135,21 @@ Feature: SetMessages method on shared folders
     Then "bob@domain.tld" ask for message "mDraft"
     And the notFound list should contain "mDraft"
 
-  Scenario: Draft creation in outbox is not allowed
+  Scenario: Draft creation in outbox is allowed
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature]: I don't think so, It should be the only case forbidden

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java [hunk 3]
```diff
@@ -139,7 +139,7 @@ private void update(MessageId messageId, UpdateMessagePatch updateMessagePatch,
         } catch (InvalidOutboxMoveException e) {
             ValidationResult invalidPropertyMailboxIds = ValidationResult.builder()
                 .property(MessageProperties.MessageProperty.mailboxIds.asFieldName())
-                .message("only drafts can be moved to Outbox")
+                .message(e.getMessage())
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java]: I prefer previous solution

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java [hunk 4]
```diff
@@ -180,58 +180,33 @@ private void assertUserIsSender(MailboxSession session, MailAddress sender) thro
...
-        }
-    }
-
-    private void assertNoDraftMessageOutOfDraftMailbox(List<Flags> futureFlags, boolean targetHasNonDraft) throws DraftMessageMailboxUpdateException {
-        if (targetHasNonDraft) {
-            boolean anyDraftMessage = futureFlags.stream().anyMatch(flags -> flags.contains(Flags.Flag.DRAFT));
-            if (anyDraftMessage) {
-                throw new DraftMessageMailboxUpdateException("Messages with '$Draft' keyword are prohibited in non drafts mailboxes");
-            }
+    private void assertOutboxMoveOnlyFromDraft(boolean targetIsOutbox, boolean allMessagesWhereDrafts) {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java]: method name is not ok

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 1 above)
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: You're right. I should still ensure that when we create a message in Outbox, it should be only in Outbox.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature (hunk 2 above)
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature]: Why? 

It do not violate any assumptions, as I might move a $Draft mail to Outbox without resetting $Draft flag.

Why would we care in one case, and not in the other one?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java (hunk 3 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java]: InvalidOutboxMoveException fits the scope of the two assertions, I don't want to introduce pointless overspecialised exception that don't add value.

Comment (mbaechler) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java]: Ok, don't mind

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature (hunk 2 above)
Comment (mbaechler) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature]: >  as I might move a $Draft mail to Outbox without resetting $Draft flag.

Is it allowed ? I thought it wasn't.

Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature]: The entire goal of that ticket is to not require INBOX to reset the $Draft flag while sending a draft ;-)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java (hunk 4 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java]: I propose `assertOutboxMoveOriginallyHasDraftKeywordSet`

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 5]
```diff
@@ -1499,12 +1496,9 @@ public void setMessagesShouldRejectCreateInDraftAndOutboxForASingleMessage() {
...
             .body(NAME, equalTo("messagesSet"))
-            .body(ARGUMENTS + ".created", aMapWithSize(0))
-            .body(ARGUMENTS + ".notCreated", aMapWithSize(1))
-            .body(ARGUMENTS + ".notCreated", hasKey(messageCreationId))
-            .body(ARGUMENTS + ".notCreated[\"" + messageCreationId + "\"].type", equalTo("invalidProperties"))
-            .body(ARGUMENTS + ".notCreated[\"" + messageCreationId + "\"].properties", contains("mailboxIds"))
-            .body(ARGUMENTS + ".notCreated[\"" + messageCreationId + "\"].description", equalTo("Message creation is only supported in mailboxes with role Draft and Outbox"));
+            .body(ARGUMENTS + ".notCreated", aMapWithSize(0))
+            .body(ARGUMENTS + ".created", aMapWithSize(1))
+            .body(ARGUMENTS + ".created", hasKey(messageCreationId));
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: the test name needs to be updated

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java [hunk 6]
```diff
@@ -178,40 +178,30 @@ private void handleCreate(CreationMessageEntry create, Builder responseBuilder,
     }
     
     private void performCreate(CreationMessageEntry entry, Builder responseBuilder, MailboxSession session) throws MailboxException, InvalidMailboxForCreationException, MessagingException, AttachmentsNotFoundException {
-        if (isAppendToMailboxWithRole(Role.DRAFTS, entry.getValue(), session)) {
+        if (isDraft(entry.getValue())) {
             saveDraft(entry, responseBuilder, session);
         } else if (isAppendToMailboxWithRole(Role.OUTBOX, entry.getValue(), session)) {
             sendMailViaOutbox(entry, responseBuilder, session);
         } else {
+            if (isAppendToMailboxWithRole(Role.DRAFTS, entry.getValue(), session)) {
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java]: do we need to keep this?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java [hunk 7]
```diff
@@ -180,58 +180,33 @@ private void assertUserIsSender(MailboxSession session, MailAddress sender) thro
...
             .map(MessageResult::getMailboxId)
             .collect(Guavate.toImmutableList());
-        List<Flags> futureFlags = patchFlags(messagesToBeUpdated, updateMessagePatch);
         List<MailboxId> targetMailboxes = getTargetedMailboxes(previousMailboxes, updateMessagePatch);
 
-        boolean originIsDraft = previousMailboxes.stream().allMatch(draftMailboxes::contains);
-        boolean targetIsOutbox = targetMailboxes.stream().anyMatch(outboxMailboxes::contains);
-        boolean targetHasDraft = targetMailboxes.stream().anyMatch(draftMailboxes::contains);
-        boolean targetHasNonDraft = targetMailboxes.stream().anyMatch(id -> !draftMailboxes.contains(id));
+        boolean allMessagesWhereDrafts = messagesToBeUpdated.stream()
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java]: Where?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 8]
```diff
@@ -1496,9 +1496,12 @@ public void setMessagesShouldRejectCreateInDraftAndOutboxForASingleMessage() {
             .log().ifValidationFails()
             .statusCode(200)
             .body(NAME, equalTo("messagesSet"))
-            .body(ARGUMENTS + ".notCreated", aMapWithSize(0))
-            .body(ARGUMENTS + ".created", aMapWithSize(1))
-            .body(ARGUMENTS + ".created", hasKey(messageCreationId));
+            .body(ARGUMENTS + ".created", aMapWithSize(0))
+            .body(ARGUMENTS + ".notCreated", aMapWithSize(1))
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: oh and now the test name is ok :)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java (hunk 7 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java]: Were

**Comment by chibenwa** [comment]
test this please

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java (hunk 6 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java]: Yes we need to keep this: https://james.open-paas.org/job/workflow/274/console

The idea is to provide an accurate error message when the intent of the user is to save a draft.

Scenario : Saving a message in Draft mailbox without $Draft flag
Expected error: keywords: "Expecting draft messages to have $Draft keyword"
Without that code: mailboxIds invalid: "The only implemented feature is sending via outbox and draft saving"

I will re-add it.

**Comment by aduprat** [comment]
_git push -f_

**Pull request merged**